src/data/fin/basic.lean

```diff
@@ -1617,6 +1617,47 @@ begin
     { rwa [cast_succ_pred_eq_pred_cast_succ , fin.pred_le_pred_iff] } }
 end
 
+/-- `succ_above` as an order isomorphism between `fin n` and `{x : fin (n + 1) // x ≠ p}`. -/
+def succ_above_equiv (p : fin (n + 1)) : fin n ≃o {x : fin (n + 1) // x ≠ p} :=
```

Comment:
See also `fin_succ_equiv'` in `logic.equiv.fin`.

Reply:
It's related, certainly, but it's not clear to me that it would help that much in constructing this equivalence to a subtype.

Reply:
I've got an alternative version working based on

```lean
def option_subtype [decidable_eq β] (x : β) :
  {e : option α ≃ β // e none = x} ≃ (α ≃ {y : β // y ≠ x}) :=
```

which then allows a simpler definition in terms of `fin_succ_equiv'`

```lean
def fin_succ_above_equiv (p : fin (n + 1)) : fin n ≃o {x : fin (n + 1) // x ≠ p} :=
{ map_rel_iff' := p.succ_above.map_rel_iff',
  ..equiv.option_subtype p ⟨(fin_succ_equiv' p).symm, rfl⟩ }
```

(but doesn't simplify proofs, and more proofs are needed to set up the API).

Should I PR `equiv.option_subtype` and then replace the definition in this PR with the one using `option_subtype`?

Reply:
I've opened #16302 with the `equiv.option_subtype` definition that would allow reworking this PR in terms of `fin_succ_equiv'`.

Reply:
I made this PR depend on #16302 (which is on the queue)

Reply:
I've now replaced the implementation in this PR with one based on `equiv.option_subtype` and `fin_succ_equiv'`.